### PR TITLE
Newsletter bulk actions refactor + deleting fix [MAILPOET-2898][MAILPOET-2758]

### PR DESCRIPTION
**Pull request by costasovo**
[MAILPOET-2898][MAILPOET-2758]

[MAILPOET-2898]: https://mailpoet.atlassian.net/browse/MAILPOET-2898
[MAILPOET-2758]: https://mailpoet.atlassian.net/browse/MAILPOET-2758

### CODE REVIEW

File: lib/API/JSON/v1/Newsletters.php [hunk 1]
```diff
@@ -202,15 +202,12 @@ public function setStatus($data = []) {
...
   public function restore($data = []) {
     $id = (isset($data['id']) ? (int)$data['id'] : false);
-    $newsletter = Newsletter::findOne($id);
-    if ($newsletter instanceof Newsletter) {
-      $newsletter->restore();
-
-      $newsletter = Newsletter::findOne($newsletter->id);
-      if(!$newsletter instanceof Newsletter) return $this->errorResponse();
-
+    $newsletter = $this->newslettersRepository->findOneById($id);
```
Comment (JanJakes) [inline lib/API/JSON/v1/Newsletters.php:205]: Use `$this->getNewsletter($data);`.

File: lib/API/JSON/v1/Newsletters.php [hunk 2]
```diff
@@ -222,14 +219,12 @@ public function restore($data = []) {
 
   public function trash($data = []) {
     $id = (isset($data['id']) ? (int)$data['id'] : false);
-    $newsletter = Newsletter::findOne($id);
-    if ($newsletter instanceof Newsletter) {
-      $newsletter->trash();
-
-      $newsletter = Newsletter::findOne($newsletter->id);
-      if(!$newsletter instanceof Newsletter) return $this->errorResponse();
+    $newsletter = $this->newslettersRepository->findOneById($id);
```
Comment (JanJakes) [inline lib/API/JSON/v1/Newsletters.php:222]: Use `$this->getNewsletter($data);`.

File: lib/API/JSON/v1/Newsletters.php [hunk 3]
```diff
@@ -241,9 +236,9 @@ public function trash($data = []) {
 
   public function delete($data = []) {
     $id = (isset($data['id']) ? (int)$data['id'] : false);
-    $newsletter = Newsletter::findOne($id);
-    if ($newsletter instanceof Newsletter) {
-      $newsletter->delete();
+    $newsletter = $this->newslettersRepository->findOneById($id);
```
Comment (JanJakes) [inline lib/API/JSON/v1/Newsletters.php:239]: Use `$this->getNewsletter($data);`.

File: lib/Newsletter/NewslettersRepository.php [hunk 4]
```diff
@@ -126,6 +126,15 @@ public function bulkTrash(array $ids): int {
     if (empty($ids)) {
       return 0;
     }
+    // Fetch children id for trashing
+    $childrenIds = $this->entityManager->createQueryBuilder()->select( 'n.id')
```
Comment (JanJakes) [inline lib/Newsletter/NewslettersRepository.php:130]: Extra space: `select( 'n.id')`

File: lib/Newsletter/NewslettersRepository.php [hunk 5]
```diff
@@ -126,6 +126,15 @@ public function bulkTrash(array $ids): int {
...
+      ->from(NewsletterEntity::class, 'n')
+      ->where('n.parent IN (:ids)')
+      ->setParameter('ids', $ids)
+      ->getQuery()->getScalarResult();
```
Comment (JanJakes) [inline lib/Newsletter/NewslettersRepository.php:134]: "Fetch children IDs for <action>" is always the same, right? Could be a private method.

File: lib/Newsletter/NewslettersRepository.php [hunk 6]
```diff
@@ -104,4 +118,211 @@ public function getAnalytics(): array {
...
+         DELETE sn FROM $statsNotificationsTable sn
+         WHERE sn.`newsletter_id` IN (:ids)
+      ", ['ids' => $ids], ['ids' => Connection::PARAM_INT_ARRAY]);
+
+      // Delete scheduled tasks and scheduled task subscribers
+      $scheduledTasksTable = $this->entityManager->getClassMetadata(ScheduledTaskEntity::class)->getTableName();
+      $sendingQueueTable = $this->entityManager->getClassMetadata(SendingQueueEntity::class)->getTableName();
+      $scheduledTaskSubscribersTable = $this->entityManager->getClassMetadata(ScheduledTaskSubscriberEntity::class)->getTableName();
+
+      // Delete sending tasks subscribers
```
Comment (JanJakes) [inline lib/Newsletter/NewslettersRepository.php:289]: I'd change "sending" -> "scheduled" to avoid confusion.

**Comment by JanJakes** [comment]
One more thing - I believe we should be able to "revert this revert": https://github.com/mailpoet/mailpoet/commit/1d7fb55ae42071ee338057b393bfca52989bd80a

(= Apply again this cleanup https://github.com/mailpoet/mailpoet/commit/99f4a377c7b43c999ce8c9ed62f2bca8c7bb1331.)

**Comment by costasovo** [comment]
@JanJakes Thanks for comments. I've addressed them all.

**Pull request merged**